src/Model/LazyBSONArray.php

```diff
@@ -0,0 +1,276 @@
+<?php
+/*
+ * Copyright 2023-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace MongoDB\Model;
+
+use AppendIterator;
+use ArrayAccess;
+use ArrayIterator;
+use CallbackFilterIterator;
+use IteratorAggregate;
+use MongoDB\BSON\PackedArray;
+use MongoDB\Codec\CodecLibrary;
+use MongoDB\Codec\LazyBSONCodecLibrary;
+use MongoDB\Exception\InvalidArgumentException;
+use ReturnTypeWillChange;
+
+use function array_key_exists;
+use function array_keys;
+use function array_map;
+use function array_merge;
+use function array_values;
+use function is_array;
+use function is_numeric;
+use function max;
+use function MongoDB\recursive_copy;
+use function sprintf;
+use function trigger_error;
+
+use const E_USER_WARNING;
+
+/**
+ * Model class for a BSON array.
+ *
+ * The internal data will be filtered through array_values() during BSON
+ * serialization to ensure that it becomes a BSON array.
+ *
+ * @template TValue
+ * @template-implements ArrayAccess<int, TValue>
+ * @template-implements IteratorAggregate<int, TValue>
+ */
+class LazyBSONArray implements ArrayAccess, IteratorAggregate
```

Comment:
Implemented `Countable` as well as magic methods for serialisation. I'm not sure if we need `JsonSerializable`, as that would technically mean serialising to BSON first and then converting that BSON to either extended or relaxed canonical JSON format.

Reply:
To implement a simple JSON HTTP API, one would need to use `json_encode` on the documents. I don't understand what you means by "serialising to BSON first and then converting that BSON to either extended or relaxed canonical JSON format."

```php
public function searchAction($terms): JsonResponse
{
    /** @var LazyBSONArray.php */
    $documents = $this->repository->search($terms);

    return new JsonResponse($documents); // uses json_encode($documents)
}
```

It's not that important. It can be implemented later if the need arises.

Reply:
Good point. I've implemented it by returning the iterator as array, meaning that data is decoded from BSON. I misunderstand the original purpose.